File: src/main/java/greedy/WordBreak.java
package greedy;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Scanner;

/**
 * Created by tpusers on 2017/2/19.
 */
public class WordBreak {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String s = scanner.next();

        String[] strings = scanner.next().split(",");
        List<String> dict = Arrays.asList(strings);
        System.out.println(wordBreak(s, dict));
    }
    public static Boolean wordBreak(String s, List<String> dict) {
        boolean[] f = new boolean[s.length() + 1];
        f[0] = true;
        for (int i = 1; i <= s.length(); i ++)
            for (int j = 0; j < i; j ++)
                if (f[j] && dict.contains(s.substring(j, i))){
                    f[i] = true;
                    break;
                }
        return f[s.length()];
    }
}

File: src/main/scala/Greedy/WordBreak.scala
package Greedy

import scala.io.Source

/**
  * Created by tpusers on 2017/2/19.
  */
object WordBreak {
  def main(args: Array[String]): Unit = {
    val s = Source.stdin.getLines()
    val str = s.next()
    val dict = s.next().split(',').toList
    print(wordBreak(str, dict))
  }
  def wordBreak(s: String, dict: List[String]) = {
    val f = new Array[Boolean](s.length + 1)
    f(0) = true

    for (i <- 1 to s.length){
      for (j <- 0 until i)
        if (f(j) && dict.contains(s.substring(j, i)))
          f(i) = true
    }
    f(s.length)
  }

}
